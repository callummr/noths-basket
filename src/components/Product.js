import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const ProductCard = styled.article`
  @media (min-width: 415px) {
    width: 45%;
  }
  @media (min-width: 768px) {
    width: 30%;
  }
  width: 100%;
  margin-bottom: 20px;
  background-color: #fff;
  padding: 15px;
  position: relative;
  padding-bottom: 70px;
`

const Image = styled.img`
  width: 100%;
  margin-bottom: 10px;
`

const Title = styled.h1`
  color: #333333;
  font-size: 19px;
  margin-bottom: 15px;
  line-height: 1.2;
`

const Price = styled.span`
  color: green;
`

const Button = styled.button`
  position: absolute;
  height: 40px;
  background-color: #0a89af;
  bottom: 15px;
  left: 15px;
  right: 15px;
  line-height: 40px;
  text-align: center;
  text-decoration: none;
  font-size: 15px;
  color: #fff;
  border-radius: 5px;
  font-weight: 600;
  width: calc(100% - 30px);
  padding: 0;

  &:hover {
    opacity: 0.8;
  }
`

const Product = props => (
  <ProductCard>
    <Image
      src={props.imageUrl || 'http://via.placeholder.com/150x150'}
      alt={props.name}
    />
    <Title>{props.name}</Title>
    <Price>Â£{props.price.toFixed(2)}</Price>
    <Button onClick={() => props.onAddToBasket(props.id)}>Add to Basket</Button>
  </ProductCard>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  imageUrl: PropTypes.string,
  price: PropTypes.number.isRequired,
  onAddToBasket: PropTypes.func.isRequired,
}

export default Product
