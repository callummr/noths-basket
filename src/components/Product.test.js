import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Product from './Product'

const productProps = {
  id: '001',
  name: 'test-name',
  imageUrl: 'http://test.image',
  price: 1.5,
  onAddToBasket: jest.fn(),
}

describe('Product', () => {
  beforeEach(() => {
    productProps.onAddToBasket.mockClear()
  })

  it('matches snapshot', () => {
    const instance = <Product {...productProps} />
    expect(renderer.create(instance).toJSON()).toMatchSnapshot()
  })

  it('renders image if URL given', () => {
    const wrapper = mount(<Product {...productProps} />)
    expect(wrapper.find('img').prop('src')).toBe(productProps.imageUrl)
  })

  it('renders a placeholder if no imageURL given', () => {
    const wrapper = mount(
      <Product {...{ ...productProps, imageUrl: undefined }} />
    )
    expect(
      wrapper
        .find('img')
        .prop('src')
        .includes('placeholder')
    ).toBe(true)
  })

  it('calls onAddToBasket when button clicked', () => {
    const wrapper = mount(<Product {...productProps} />)
    wrapper.find('button').simulate('click')
    expect(productProps.onAddToBasket).toHaveBeenCalled()
  })
})
