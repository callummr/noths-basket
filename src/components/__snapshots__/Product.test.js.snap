// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product matches snapshot 1`] = `
<article
  className="sc-bdVaJa bUUZdf"
>
  <img
    alt="test-name"
    className="sc-bwzfXH qVTvB"
    src="http://test.image"
  />
  <h1
    className="sc-htpNat lloKWr"
  >
    test-name
  </h1>
  <span
    className="sc-bxivhb jiHAbL"
  >
    Â£
    1.50
  </span>
  <button
    className="sc-ifAKCX cpSZLT"
    onClick={[Function]}
  >
    Add to Basket
  </button>
</article>
`;
